### PR TITLE
Workaround for nonnull operator on indexed accesses

### LINKED ISSUE #17005: Readonly for generic param doesn't work for optional properties
**TypeScript Version:**  2.3.4 or 2.4.1

**Code**

```ts
interface A {
    params?: { name: string; }
}

class Test<T extends A> {
    attrs: Readonly<T>;

    m() {
        this.attrs.params!.name;                    
    }
}
```

**Expected behavior:**
no errors

**Actual behavior:**
Error: Property 'name' does not exist on type 'T["params"]'.

If I remove `Readonly` or get rid of generic param `T` then everything will be ok.

### PR DESCRIPTION
**Pull request by weswigham**
We now force the substitution of an indexed access with its base constraint when getting a type with facts, allowing the resulting type to be filtered. The correct approach would be to apply a lazy subtraction type to the indexed access type, and defer the evaluation until required; however @rbuckton is already working on something which can subsume that purpose.

Fixes #17005

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11378,6 +11378,18 @@ namespace ts {
         }
 
         function getTypeWithFacts(type: Type, include: TypeFacts) {
+            if (type.flags & TypeFlags.IndexedAccess) {
+                // TODO (weswig): This is a substitute for a lazy negated type to remove the types indicated by the TypeFacts from the (potential) union the IndexedAccess refers to
+                //  - instead of defering the resolution of the access, we apply the facts to the base constraint of the access instead; so this works under most circumstances,
+                //    like when the index refers to an optional member and you want to access that member, but unfortunately if that member's type is supposed to vary with the base,
+                //    then the eager evaluation of this removes that relationship
```
Comment (sandersn) [inline src/compiler/checker.ts]: 1. Can you add a short comment to the test that links the test case to this limitation?
2. Can you put in an example here?
3. Can you use fewer words if you have an example?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -11378,6 +11378,18 @@ namespace ts {
         }
 
         function getTypeWithFacts(type: Type, include: TypeFacts) {
+            if (type.flags & TypeFlags.IndexedAccess) {
+                // TODO (weswig): This is a substitute for a lazy negated type to remove the types indicated by the TypeFacts from the (potential) union the IndexedAccess refers to
+                //  - instead of defering the resolution of the access, we apply the facts to the base constraint of the access instead; so this works under most circumstances,
+                //    like when the index refers to an optional member and you want to access that member, but unfortunately if that member's type is supposed to vary with the base,
+                //    then the eager evaluation of this removes that relationship
+                const innerType = getBaseConstraintOfType(type) || emptyObjectType;
```
Comment (sandersn) [inline src/compiler/checker.ts]: Name: constraint or baseConstraint?

**Comment by weswigham** [comment]
> I'm having trouble with an example of the case that this hack breaks. Can you explain more? (See below for details)

No so much breaks as does not fully fix, for example in:
```
interface A {
    params?: { name: string; }
}

class Test<T extends A> {
    attrs: Readonly<T>;

    m() {
        return this.attrs.params!;                    
    }
}
```
The return type of `m` should be `NonNull<T["params"]>` - right now it is `T["params"]` (no null/undefined removed), and with this change it is `{name: string}` (which is no longer associated with `T`). Previously if a subclass gave a more specific `"params"` member type, the `m` method would return that more specific type - but now not so much.

**Comment by weswigham** [comment]
@sandersn Got your comments. The bulk of the `TODO` comment has been replaced with a reference to this PR and the associated test, and the associated test now has an example showing where the current behavior is undesirable.

**Comment by weswigham** [comment]
@mmitche The dotnet CI builds on this PR seem to be using an outdated node modules cache (build is failing because it's being build against `@types/node@4` instead of `8`) - any idea why?

**Comment by mmitche** [comment]
@weswigham No idea...I don't know much about node.

**Pull request merged**